### PR TITLE
Replace WeakReference with WeakReference<T>

### PR DESCRIPTION
**Pull request by Wraith2**
Follow-up from https://github.com/dotnet/SqlClient/pull/1122 , I checked for more uses of `WeakReference` and replaced them with `WeakReference<T>`.

### CODE REVIEW

**Review by johnnypham — APPROVED** [review]

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/sqlinternaltransaction.cs [hunk 1]
```diff
@@ -591,20 +590,16 @@ internal void Zombie()
 
         private void ZombieParent()
         {
-            if (null != _parent)
+            if (_parent != null && _parent.TryGetTarget(out SqlTransaction parent))
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/sqlinternaltransaction.cs:593]: As per current design, should we ensure `_parent` is set to `null` even if `_parent.TryGetTarget` returns false?

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlInternalTransaction.cs [hunk 2]
```diff
@@ -533,20 +533,16 @@ internal void Zombie()
 
         private void ZombieParent()
         {
-            if (null != _parent)
+            if (_parent != null && _parent.TryGetTarget(out SqlTransaction parent))
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlInternalTransaction.cs:536]: Same here

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/sqlinternaltransaction.cs (hunk 1 above)
Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/sqlinternaltransaction.cs:593]: if parent is null TryGetTarget will return false, I checked the code in runtime. The corollary is also true, if TryGetTarget returns true then parent is guaranteed not to be null.

Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/sqlinternaltransaction.cs:593]: I meant reverse, if **parent** is **not** null, will **TryGetTarget** always return **true**?
If no, then we have a case where parent will not be zombied to null, if there was no target attached.

Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/sqlinternaltransaction.cs:593]: > if parent is not null, will TryGetTarget always return true?
yes.

Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/sqlinternaltransaction.cs:593]:
```csharp
        public bool TryGetTarget([MaybeNullWhen(false), NotNullWhen(true)] out T target)
        {
            // Call the worker method that has more performant but less user friendly signature.
            T? o = this.Target;
            target = o!;
            return o != null;
        }
```

https://github.com/dotnet/runtime/blob/efd818b90d61a81b73b20f3d1720279f993faa2c/src/libraries/System.Private.CoreLib/src/System/WeakReference.T.cs#L54-L60

You can tell from the naming that is is old, o wouldn't pass api review today.

Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/sqlinternaltransaction.cs:593]: Well, I'd say it depends on whether there was an item added to it, but is not going to return "true" _by design_. Just to keep things consistent with current design, I'd recommend putting it in second block and setting parent to _null_ never-the-less.

Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/sqlinternaltransaction.cs:593]: Done.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlInternalTransaction.cs (hunk 2 above)
Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlInternalTransaction.cs:536]: Done.

**Review by cheenamalhotra — APPROVED** [review]

**Pull request merged**